### PR TITLE
Sasl external best practices

### PR DESCRIPTION
**Pull request by zofpolkowska**
This PR addresses best practices for using `SASL EXTERNAL` as defined by https://xmpp.org/extensions/xep-0178.html.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[6137.1](https://travis-ci.org/esl/MongooseIM/jobs/485921797) / Erlang 19.3 / small_tests / de8feecd686abefe034fce07a428e3fb6c128a0f
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2204/6137/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2204/6137/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-29_14.18.21/lib.mongooseim.logs/run.2019-01-29_14.18.22/suite.log.html)

**Comment by mongoose-im** [comment]
[6138.1](https://travis-ci.org/esl/MongooseIM/jobs/485924853) / Erlang 19.3 / small_tests / 2b6c15208d968f540e4d317460dda552be78e630
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2204/6138/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2204/6138/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-29_14.55.29/lib.mongooseim.logs/run.2019-01-29_14.55.31/suite.log.html)

File: src/sasl/cyrsasl_external.erl [hunk 1]
```diff
@@ -49,21 +49,41 @@ mech_new(_Host, Creds) ->
...
+    case mongoose_credentials:get(Creds, xmpp_addresses) of
+        [_OneXmppAddr] ->
+            authorize(Creds);
+        [] ->
+            authorize(Creds);
+        _ ->
+            {error, <<"invalid-authzid">>}
+    end;
+do_mech_step(#state{creds = Creds}, User) ->
+    case mongoose_credentials:get(Creds, xmpp_addresses) of
```
Comment (DenysGonchar) [inline src/sasl/cyrsasl_external.erl]: according to XEP user name must be provided if cert has more than one item in xmpp_address field. seems like this case clause is missing.

Comment (zofpolkowska) [reply src/sasl/cyrsasl_external.erl]: Thanks! I added it, then removed it and now re-added :)

**Comment by mongoose-im** [comment]
[6143.1](https://travis-ci.org/esl/MongooseIM/jobs/486361495) / Erlang 19.3 / small_tests / a3fad8a31f78e918d4f0c25abc4db75bd3fc5555
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2204/6143/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2204/6143/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-30_11.16.10/lib.mongooseim.logs/run.2019-01-30_11.16.11/suite.log.html)

**Comment by mongoose-im** [comment]
[6145.1](https://travis-ci.org/esl/MongooseIM/jobs/486454022) / Erlang 19.3 / small_tests / 8718903bab2bc140c7792eac866fbd2dbd3ad341
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2204/6145/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2204/6145/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-30_14.53.17/lib.mongooseim.logs/run.2019-01-30_14.53.18/suite.log.html)

File: big_tests/tests/sasl_external_SUITE.erl [hunk 2]
```diff
@@ -191,7 +235,8 @@ no_cert_fails_to_authenticate(_C) ->
 
 generate_certs(C) ->
     Certs = [{maps:get(cn, CertSpec), generate_cert(C, CertSpec)} ||
-	     CertSpec <- [#{cn => "not-alice-name", xmpp_addrs => ["alice@localhost", "alice@fed1"]},
+	     CertSpec <-
+             [#{cn => "not-alice-name", xmpp_addrs => ["alice@localhost", "alice@fed1"]},
```
Comment (michalwski) [inline big_tests/tests/sasl_external_SUITE.erl]: The indentation inside the list is broken now.

File: src/sasl/cyrsasl_external.erl [hunk 3]
```diff
@@ -49,28 +50,63 @@ mech_new(_Host, Creds) ->
...
+    case mongoose_credentials:get(Creds, cert_file) of
+        false ->
+            {error, <<"not-authorized">>};
+        true ->
+            do_mech_step(Creds, User)
+    end.
+do_mech_step(Creds, User) ->
+    XmppAddrs = mongoose_credentials:get(Creds, xmpp_addresses),
+    CommonName = mongoose_credentials:get(Creds, common_name),
+    case authorize(XmppAddrs, CommonName, User) of
```
Comment (michalwski) [inline src/sasl/cyrsasl_external.erl]: I think the name of the function could be changed to sth like `check_auth_request` and return `{ok, Username}` or `{error, Error}`.

**Comment by mongoose-im** [comment]
[6147.5](https://travis-ci.org/esl/MongooseIM/jobs/486473574) / Erlang 19.3 / ldap_mnesia / 273e4418545128a139261cf5eabda1415be1f0d0
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2204/6147/ldap_mnesia.19.3)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2204/6147/ldap_mnesia.19.3/big/ct_run.test%40travis-job-97d99290-283c-401a-9eb3-f8e0087ef2e7.2019-01-30_15.40.27/index.html)
**OK: 1170** / **Failed: 1** / **User-skipped: 99** / Auto-skipped: 0 
<details><summary>mod_global_distrib_SUITE:mod_global_distrib:test_pm_with_ungraceful_reconnection_to_different_server</summary>


```erlang
{error,
  {timeout_when_waiting_for_stanza,
    [{escalus_client,wait_for_stanza,
       [{client,<<"eve42.33960@localhost/res1">>,escalus_tcp,
          <0.14799.1>,
          [{event_manager,<0.14790.1>},
           {server,<<"localhost">>},
           {username,<<"eve42.33960">>},
           {resource,<<"res1">>}],
          [{event_client,
             [{event_manager,<0.14790.1>},
            {server,<<"localhost">>},
            {username,<<"eve42.33960">>},
            {resource,<<"res1">>}]},
           {resource,<<"res1">>},
           {username,<<"eve42.33960">>},
           {server,<<"localhost">>},
           {host,<<"localhost">>},
           {port,5222},
           {auth,{escalus_auth,auth_plain}},
           {wspath,undefined},
           {username,<<"eve42.33960">>},
           {server,<<"localhost">>},
           {password,<<"password">>},
           {port,5222},
           {stream_management,true},
           {stream_id,<<"D4F4B8782409CE58">>}]},
        10000],
       [{file,
          "/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_client.erl"},
        {line,138}]},
     {mod_global_distrib_SUITE,
       '-test_pm_with_ungraceful_reconnection_to_different_server/1-fun-0-',
       4,
       [{file,"mod_global_distrib_SUITE.erl"},{line,607}]},
     {escalus_story,story,4,
       [{file,
          "/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
        {line,72}]},
     {tes...
```

[Report log](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2204/6147/ldap_mnesia.19.3/big/ct_run.test%40travis-job-97d99290-283c-401a-9eb3-f8e0087ef2e7.2019-01-30_15.40.27/big_tests.tests.mod_global_distrib_SUITE.logs/run.2019-01-30_15.53.54/mod_global_distrib_suite.test_pm_with_ungraceful_reconnection_to_different_server.html)
</details>

**Comment by mongoose-im** [comment]
[6148.1](https://travis-ci.org/esl/MongooseIM/jobs/486821262) / Erlang 19.3 / small_tests / 475026f422a8b5474d49f4b46c2051b62c275443
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2204/6148/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2204/6148/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-31_08.38.14/lib.mongooseim.logs/run.2019-01-31_08.38.15/suite.log.html)

**Comment by mongoose-im** [comment]
[6152.1](https://travis-ci.org/esl/MongooseIM/jobs/486873423) / Erlang 19.3 / small_tests / a11fb6fa948533ef9cc4487a0875aca2b999ad89
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2204/6152/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2204/6152/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-31_11.05.31/lib.mongooseim.logs/run.2019-01-31_11.05.32/suite.log.html)

File: src/sasl/cyrsasl_external.erl [hunk 4]
```diff
@@ -45,9 +45,10 @@ stop() ->
 
 -spec mech_new(Host :: ejabberd:server(),
                Creds :: mongoose_credentials:t()) -> {ok, sasl_external_state()}.
-mech_new(_Host, Creds) ->
+mech_new(Host, Creds) ->
     Cert = mongoose_credentials:get(Creds, client_cert, no_cert),
-    maybe_extract_certs(Cert, Creds).
+    NewCreds = mongoose_credentials:extend(Creds, [{server, Host}]),
```
Comment (michalwski) [inline src/sasl/cyrsasl_external.erl]: I'm pretty sure that the information is already in Creds, did you check it? You can read it with `mongoose_credentials:lserver/1`.

File: src/sasl/cyrsasl_external.erl [hunk 5]
```diff
@@ -124,8 +126,12 @@ get_xmpp_addresses(Cert) ->
...
     mongoose_credentials:get(Cred, Key, undefined).
 
-get_username(<<"">>) ->
-    <<"">>;
-get_username(Jid) ->
+verify_server(Jid, Server) ->
     JidRecord = jid:binary_to_bare(Jid),
-    JidRecord#jid.user.
+    RequestServer = JidRecord#jid.server,
+    case RequestServer of
```
Comment (michalwski) [inline src/sasl/cyrsasl_external.erl]: I think `JidRecord#jid.server` could be used inside the `case ... of`. Also it's safer to use `lserver` as this field is normalized according to XMPP specs.

File: src/sasl/cyrsasl_external.erl [hunk 6]
```diff
@@ -44,33 +45,70 @@ stop() ->
...
+    case verify_server(User, Server) of
+        {ok, CommonName} ->
+            {ok, CommonName};
+        _ ->
+            {error, <<"not-authorized">>}
+    end;
+check_auth_req([_], _,  _, _) ->
+    {error, <<"invalid-authzid">>};
+check_auth_req(XmppAddrs, _,  User, Server) ->
+    case lists:filter(fun(XmppAddr) -> XmppAddr == User end, XmppAddrs) of
```
Comment (michalwski) [inline src/sasl/cyrsasl_external.erl]: Could `lists:member` be used here?

**Comment by mongoose-im** [comment]
[6153.1](https://travis-ci.org/esl/MongooseIM/jobs/486883164) / Erlang 19.3 / small_tests / 7382e499d6cedda1b4a656ad1b225d1770eb9230
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2204/6153/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2204/6153/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-31_11.33.06/lib.mongooseim.logs/run.2019-01-31_11.33.07/suite.log.html)

**Comment by mongoose-im** [comment]
[6156.1](https://travis-ci.org/esl/MongooseIM/jobs/486895625) / Erlang 19.3 / small_tests / d3f2ae0caf98eed63a1c7bccbea1b2f55532ebd6
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2204/6156/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2204/6156/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-31_11.57.15/lib.mongooseim.logs/run.2019-01-31_11.57.16/suite.log.html)

File: doc/authentication-methods/client-certificate.md [hunk 7]
```diff
@@ -28,6 +28,13 @@ A `SASL EXTERNAL` authentication method is disabled by default.
 In order to enable it, please add [`sasl_mechanisms` option](../Advanced-configuration.md#authentication) to MongooseIM config file.
 Its value must include a `cyrsasl_external` item.
 Obviously the list may be longer, if the system should support both certificate and password based authentication.
```
Comment (goddammit) [inline doc/authentication-methods/client-certificate.md]: Obviously the list may be longer, if the system should support both the certificate and password based authentication.

File: doc/authentication-methods/client-certificate.md [hunk 8]
```diff
@@ -28,6 +28,13 @@ A `SASL EXTERNAL` authentication method is disabled by default.
...
+`SASL EXTERNAL` authentication method requires the use of digital
```
Comment (goddammit) [inline doc/authentication-methods/client-certificate.md]: The `SASL EXTERNAL` authentication method requires the use of digital certificates. 
alt:
The `SASL EXTERNAL` authentication method requires using the digital certificates.

File: doc/authentication-methods/client-certificate.md [hunk 9]
```diff
@@ -28,6 +28,13 @@ A `SASL EXTERNAL` authentication method is disabled by default.
...
+certificates. If the `xmpp_addrs` fields are included in certificate, they are always checked first.
```
Comment (goddammit) [inline doc/authentication-methods/client-certificate.md]: If the `xmpp_addrs` fields are included in the certificate, they are always checked first.

File: doc/authentication-methods/client-certificate.md [hunk 10]
```diff
@@ -28,6 +28,13 @@ A `SASL EXTERNAL` authentication method is disabled by default.
...
+In case no `xmpp_addrs` are specified, the `cn` (_common name_) field might be
```
Comment (goddammit) [inline doc/authentication-methods/client-certificate.md]: When no `xmpp_addrs` are specified, the `cn` (_common name_) field might be used, but it is optional.

File: doc/authentication-methods/client-certificate.md [hunk 11]
```diff
@@ -28,6 +28,13 @@ A `SASL EXTERNAL` authentication method is disabled by default.
...
+used, but it is optional. If there are more than one `xmpp_addrs` or the
```
Comment (goddammit) [inline doc/authentication-methods/client-certificate.md]: "Or" or "on"?
If or:
If there is more than one `xmpp_addrs` or both the list and `cn` field are empty, the client must include the authorization entity.

if on:

If there is more than one `xmpp_addrs` on the list and the`cn` field is empty, the client must include the authorization entity.

File: src/sasl/cyrsasl_external.erl [hunk 12]
```diff
@@ -44,33 +45,69 @@ stop() ->
...
+do_mech_step(Creds, User) ->
+    XmppAddrs = get_credentials(Creds, xmpp_addresses),
+    CommonName = get_credentials(Creds, common_name),
+    Server = mongoose_credentials:lserver(Creds),
+    case check_auth_req(XmppAddrs, CommonName, User, Server) of
+        {error, Error} ->
+            {error, Error};
+        {ok, Name} ->
+            NewCreds = mongoose_credentials:extend(Creds, [{username, Name}]),
+            ejabberd_auth:authorize(NewCreds)
```
Comment (chrzaszcz) [inline src/sasl/cyrsasl_external.erl]: This function may return `{error, not_authorized}`, why the specs in line 63 say `{error, binary()}`.

**Pull request merged**